File: yiseungyun/week23/14501.py
import sys  
input = sys.stdin.readline

N = int(input())
counsel = []
for _ in range(N):
  T, P = map(int, input().split())
  counsel.append([T, P])

dp = [0 for _ in range(N+1)]

for i in range(N-1, -1, -1):
  if i+counsel[i][0] <= N: 
    dp[i] = max(dp[i+counsel[i][0]]+counsel[i][1], dp[i+1])
  else: 
    dp[i] = dp[i+1]

print(dp[0])

File: iam454/week23/14501.js
const fs = require("fs");

const input = fs
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "./input.txt")
  .toString()
  .trim()
  .split("\n");

const N = Number(input.shift());
const table = input.map((item) => item.split(" ").map(Number));

const dp = new Array(N + 1).fill(0);

for (let i = 0; i < N; i++) {
  const [T, P] = table[i];

  if (i + T <= N) {
    dp[i + T] = Math.max(dp[i + T], dp[i] + P);
  }

  dp[i + 1] = Math.max(dp[i + 1], dp[i]);
}

console.log(dp[N]);
